packages/react-dom/src/events/ReactEventDispatchTracker.js

```diff
@@ -0,0 +1,28 @@
+/**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {AnyNativeEvent} from 'events/PluginModuleType';
+
+const lastTarget = '__$react_event_tracking:' + Math.random();
+
+type TrackableEvent = AnyNativeEvent & {
+  [lastTarget: string]: Node,
+};
+
+export function trackEventDispatch(event: TrackableEvent) {
+  event[lastTarget] = event.currentTarget;
+}
+
+export function hasEventDispatched(event: TrackableEvent): boolean {
+  if (!event.hasOwnProperty(lastTarget)) {
+    return false;
+  }
+
+  return event[lastTarget] !== event.currentTarget;
+}
```

Comment:
I removed the WeakSet part of this when it became clear that I really needed to assign a value (and use a WeakMap). I've removed that from this module so that the intent is clearer. If there is a concern with assigning attributes to events, we can use a WeakMap to make the association, which should be supported in most browsers.